### PR TITLE
groups: support hierarchical groups.yaml files

### PR DESCRIPTION
**Pull request by nikhita**
For https://github.com/kubernetes/k8s.io/issues/460

Requiring all groups be added to a single `groups.yaml` can become
tedious to maintain. This commit allows splitting groups into
hierarchical `groups.yaml` files, akin to OWNERS files.

This commit contains the following changes:

- We now hardcode files containing groups config to have the file name
`groups.yaml`.

- We no longer allow the group config to be specified at an URL.

- The `groups-file` field in config.yaml is renamed to `groups-path`.
`groups-path` is the directory containing groups.yaml files. It must be
an absolute path. If not specified, it defaults to the directory
containing `config.yaml`.

- `readGroupsConfig` uses `groups-path` as the root dir and recursively
walks through all dirs and files. When it sees a `groups.yaml` file,
it reads the config and adds its groups to `config.Groups`.

- `config.Groups` finally contains the merged groups config from all
`groups.yaml` files.

---

This is WIP because:

- As https://github.com/kubernetes/k8s.io/issues/928#issuecomment-640993935 mentions, we probably need documented policies about how we'd split groups.yaml before enabling this. I just wanted to get this PR out to get the ball rolling.

- I moved `community@kubernetes.io` to the `sig-contributor-experience` dir and seeded OWNERS with the root OWNERS in k/community as an example, we'd probably need to tweak OWNERS/groups a bit more.

/hold
to ensure we have policies in place before enabling this

/assign @cblecker @dims @spiffxp

### CODE REVIEW

**Comment by nikhita** [comment]
cc @mrbobbytables @castrojo 
might be useful if we want to include more gsuite accounts (xref https://github.com/kubernetes/steering/issues/166)

File: groups/reconcile.go [hunk 1]
```diff
@@ -202,30 +202,66 @@ func readConfig(configFilePath string, confirmChanges bool) error {
...
+	}
+	return nil
+}
+
+// readMultipleGroupsConfig uses configFilePath's dir as the root dir and
+// recursively walks through all dirs and files. It reads the GroupsConfig
+// from all files that have the name groupsFileName and adds the groups in
+// each GroupsConfig to config.Groups.
+func readMultipleGroupsConfig(configFilePath string, groupsFileName string, config *GroupsConfig) error {
+	rootDir := filepath.Dir(configFilePath)
```
Comment (spiffxp) [inline groups/reconcile.go:230]: nit: I think it would be cleaner if this function took `rootDir` directly, and the work of deriving it was moved elsewhere

File: groups/reconcile.go [hunk 2]
```diff
@@ -202,30 +202,66 @@ func readConfig(configFilePath string, confirmChanges bool) error {
...
+
+			for _, g := range groupsConfigAtPath.Groups {
+				config.Groups = append(config.Groups, g)
+			}
 		}
+		return nil
+	})
+}
+
+func readGroupsConfigFromURL(groupsURL *url.URL, config *GroupsConfig) error {
```
Comment (spiffxp) [inline groups/reconcile.go:259]: Perhaps we should drop this functionality entirely.  It's not exercised and it's not tested.

File: groups/reconcile.go [hunk 3]
```diff
@@ -202,30 +202,66 @@ func readConfig(configFilePath string, confirmChanges bool) error {
 	return err
 }
 
-func readGroupsConfig(configFilePath string, groupsConfigFilePath string, config *GroupsConfig) error {
+func readGroupsConfig(configFilePath string, groupsFileName string, config *GroupsConfig) error {
```
Comment (spiffxp) [inline groups/reconcile.go:205]: Capitalization nit
```suggestion
func readGroupsConfig(configFilePath string, groupsFilename string, config *GroupsConfig) error {
```

As with `readMultipleGroupsConfig` I think this should take an explicit dir instead of passing along `configFilePath` and deriving from it later

File: groups/groups_test.go [hunk 4]
```diff
@@ -25,25 +25,59 @@ import (
...
 	"testing"
 	"unicode/utf8"
+
+	"k8s.io/apimachinery/pkg/util/sets"
 )
 
 var cfg GroupsConfig
 
-var groupsConfigPath = flag.String("groups-config", "./groups.yaml", "Path to groups config")
+var groupsConfigFileName = flag.String("groups-file", "groups.yaml", "Filename containing groups config")
```
Comment (spiffxp) [inline groups/groups_test.go:34]: My personal preference would be to alter things to take an explicit `groups-path` or `groups-dir`, instead of auto-deriving it from the location of `config.yaml`

I also wonder if it would be useful to allow files named something other than `groups.yaml`, in the same way that prow jobs can be defined in arbitrarily named files.  Not blocking on this, just thinking out loud. I could foresee wg-k8s-infra using this to have a rbac file, a staging groups file, etc.

Comment (nikhita) [reply groups/groups_test.go:34]: > explicit groups-path or groups-dir, instead of auto-deriving it from the location of config.yaml

> I also wonder if it would be useful to allow files named something other than groups.yaml

This is interesting! Would something like "read all yaml files under the directory specified by `groups-dir`" make sense? (I still need to look at how prow does it today)

**Comment by thockin** [comment]
Why?  Having a strong one-file-per-directory convention is nice because it
aligns with OWNERS

**Comment by thockin** [comment]
We may even want to link directory-name to group-name to lock it own to
allow OWNERS delegation more

**Comment by nikhita** [comment]
Addressed comments. PTAL. The updates are documented in the commit message and the issue body.

TLDR:

- We now hardcode files containing groups config to have the file name
`groups.yaml`.

- We no longer allow the group config to be specified at an URL.

- The `groups-file` field in config.yaml is renamed to `groups-path`.
`groups-path` is the directory containing groups.yaml files. It must be
an absolute path. If not specified, it defaults to the directory
containing `config.yaml`.

**Comment by nikhita** [comment]
> We may even want to link directory-name to group-name to lock it own to
allow OWNERS delegation more

I haven't added a validation for this because we might want to create directories based on sig names. I've added discussing the policy for splitting groups.yaml to [this week's wg-k8s-infra meeting](https://docs.google.com/document/d/16VBfsFMynA7tObzuZGPpw-sKDKfFc_T5W_E4IeEIaOQ/edit#).

File: groups/reconcile.go (hunk 2 above)
Comment (nikhita) [reply groups/reconcile.go]: > Perhaps we should drop this functionality entirely. It's not exercised and it's not tested.

Removed.

File: groups/sig-contributor-experience/OWNERS [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# See the OWNERS docs at https://go.k8s.io/owners
+
+approvers:
```
Comment (spiffxp) [inline groups/sig-contributor-experience/OWNERS:3]:
```suggestion
reviewers:
```
I would happily merge this PR as-is if this wasn't present.  I do want to open up the pool but we're not there yet.

If the goal is to fast-track membership changes, I would suggest we test against a known list of groups that is under /groups/OWNERS approval only

If the goal is to allow sigs/wgs/whatever to create whatever @kubernetes.io groups they want, I would at a minimum require adherence to common naming conventions and settings

Comment (nikhita) [reply groups/sig-contributor-experience/OWNERS:3]: Should I just removed the contribex dir and comment out the test for it? I just meant it as an example that this PR works :woman_shrugging: 

I agree that we shouldn't merge this as-is.

Comment (spiffxp) [reply groups/sig-contributor-experience/OWNERS:3]: Leaving the dir and everything is fine (and preferred!), if the approvers are changed to reviewers for now

Comment (nikhita) [reply groups/sig-contributor-experience/OWNERS]: Done!

**Comment by bartsmykla** [comment]
lgtm

**Comment by nikhita** [comment]
Discussed in today's wg k8s infra meeting, let's merge and iterate :sparkles: 

We discussed a [few approaches towards policy](https://docs.google.com/document/d/16VBfsFMynA7tObzuZGPpw-sKDKfFc_T5W_E4IeEIaOQ/edit), and decided to go with whatever is easiest to setup validation for. I'll take a dab at that over the weekend :)

/hold cancel
